_drafts/zero-to-main-2.md

```diff
@@ -0,0 +1,536 @@
+
+In order to achieve these goals, we turn to a special variable `.`, also known
+as the "location counter". The location counter tracks the current offset into a
+given memory region. As sections are added, the location counter increments
+accordingly. You can force alignemnt or gaps by setting the location counter
+forward. You may not set it backwards, and the linker will throw an error if you
+try.
+
+We set the location counter with the `ALIGN` function, to align the section, and
+use simple assignment and arithmetic to set the section size:
+
+```
+STACK_SIZE = 0x2000; /* 8 kB */
+
+SECTION {
+    ...
+    .stack (NOLOAD) :
+    {
+        . = ALIGN(8);
+        . = . + STACK_SIZE;
+        . = ALIGN(8);
+    } > ram
+    ...
+}
+```
+
+Only one to go!
+
+The `.data` section contains static variables which have an
+initial value at boot. You will remember from our previous article that since RAM
+isn't persisted while power is off, those sections need to be loaded in flash.
+At boot, the `Reset_Handler` copies the data from flash to RAM before the `main`
+function is called.
+
+To make this possible, every section in our linker script has two addresses,
+its *load* address (LMA) and its *virtual* address (VMA). In a firmware context,
+the LMA is where your JTAG loader needs to place the section and the VMA is
+where the section is found during execution.
+
+You can think of the LMA as the address "at rest" and the VMA the address during
+execution i.e. when the device is on and the program is running.
+
+The syntax to specify the LMA is relatively straightforward: every address is
+two part: <VMA> AT <LMA>. In our case it looks like this:
+
+```
+.data :
+{
+    *(.data*);
+} > ram AT > rom /* "> ram" is the VMA, "> rom" is the LMA */
+```
+
+Note that instead of appending a section to a memory region, you can explicity
+specify an address like so:
+
+```
+.data ORIGIN(ram) /* VMA */ : AT(ORIGIN(rom)) /* LMA */
+{
+    . = ALIGN(4);
+    _sdata = .;
+    *(.data*);
+    . = ALIGN(4);
+    _edata = .;
+}
+```
+
+Where `ORIGIN(<region>)` is a simple way to specify the start of a region. You
+can enter an address in hex as well.
+
+And we're done! Here's our complete linker script with every section:
+
+```
+MEMORY
+{
+  rom      (rx)  : ORIGIN = 0x00000000, LENGTH = 0x00040000
+  ram      (rwx) : ORIGIN = 0x20000000, LENGTH = 0x00008000
+}
+
+STACK_SIZE = 0x2000;
+
+/* Section Definitions */
+SECTIONS
+{
+    .text :
+    {
+        KEEP(*(.vectors .vectors.*))
+        *(.text*)
+        *(.rodata*)
+    } > rom
+
+    /* .bss section which is used for uninitialized data */
+    .bss (NOLOAD) :
+    {
+        *(.bss*)
+        *(COMMON)
+    } > ram
+
+    .data :
+    {
+        *(.data*);
+    } > ram AT >rom
+
+    /* stack section */
+    .stack (NOLOAD):
+    {
+        . = ALIGN(8);
+        . = . + STACK_SIZE;
+        . = ALIGN(8);
+    } > ram
+
+    _end = . ;
```

Comment:
curious, do you use `_end` for something or just a convention?

Reply:
Just a convention